File: 2024/day-04/main.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<string> readInput()
{
    vector<string> lines;
    string line;
    while (getline(cin, line)) {
        lines.push_back(line);
    }
    return lines;
}

bool wordIsXmas(const string &word)
{
    return word == "XMAS" || word == "SAMX";
}

bool wordIsMas(const string &word)
{
    return word == "MAS" || word == "SAM";
}

int xmasCountAt(const vector<string> &matrix, const size_t rowIndex, const size_t colIndex)
{
    const size_t numberOfRows = matrix.size();
    const size_t numberOfCols = matrix[0].size();

    int xmasCount = 0;

    // Horizontal check
    if (colIndex + 3 < numberOfCols) {
        string horizontalWord = matrix[rowIndex].substr(colIndex, 4);
        xmasCount += wordIsXmas(horizontalWord);
    }

    // Vertical check
    if (rowIndex + 3 < numberOfRows) {
        string verticalWord;
        for (size_t i = 0; i <= 3; i++) {
            verticalWord += matrix[rowIndex + i][colIndex];
        }
        xmasCount += wordIsXmas(verticalWord);
    }

    // Down-diagonal check
    if (rowIndex + 3 < numberOfRows && colIndex + 3 < numberOfCols) {
        string downDiagonalWord;
        for (size_t i = 0; i <= 3; i++) {
            downDiagonalWord += matrix[rowIndex + i][colIndex + i];
        }
        xmasCount += wordIsXmas(downDiagonalWord);
    }

    // Up-diagonal check
    if (rowIndex + 3 < numberOfRows && colIndex + 3 < numberOfCols) {
        string upDiagonalWord;
        for (size_t i = 0; i <= 3; i++) {
            upDiagonalWord += matrix[rowIndex + (3 - i)][colIndex + i];
        }
        xmasCount += wordIsXmas(upDiagonalWord);
    }

    return xmasCount;
}

bool crossMasExistsAt(const vector<string> &matrix, const size_t rowIndex, const size_t colIndex)
{
    const size_t numberOfRows = matrix.size();
    const size_t numberOfCols = matrix[0].size();

    if (rowIndex + 2 >= numberOfRows || colIndex + 2 >= numberOfCols) {
        return false;
    }

    string downDiagonalWord;
    for (size_t i = 0; i <= 2; i++) {
        downDiagonalWord += matrix[rowIndex + i][colIndex + i];
    }
    if (!wordIsMas(downDiagonalWord)) {
        return false;
    }

    string upDiagonalWord;
    for (size_t i = 0; i <= 2; i++) {
        upDiagonalWord += matrix[rowIndex + (2 - i)][colIndex + i];
    }
    if (!wordIsMas(upDiagonalWord)) {
        return false;
    }

    return true;
}

int main()
{
    const vector<string> matrix = readInput();
    const size_t numberOfRows = matrix.size();
    const size_t numberOfCols = matrix[0].size();

    int numberOfXmasOccurrences = 0;
    int numberOfCrossMasOccurrences = 0;

    for (size_t rowIndex = 0; rowIndex < numberOfRows; rowIndex++) {
        for (size_t colIndex = 0; colIndex < numberOfCols; colIndex++) {
            numberOfXmasOccurrences += xmasCountAt(matrix, rowIndex, colIndex);
            numberOfCrossMasOccurrences += crossMasExistsAt(matrix, rowIndex, colIndex);
        }
    }

    // Part one
    cout << "Number of XMAS occurrences: " << numberOfXmasOccurrences << endl;

    // Part two
    cout << "Number of X-MAS occurrences: " << numberOfCrossMasOccurrences << endl;
}

File: 2024/go/day-04/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"strings"
)

func readInput() ([][]rune, error) {
	content, err := os.ReadFile("input")
	if err != nil {
		return nil, err
	}

	lines := strings.Split(strings.TrimSpace(string(content)), "\n")
	matrix := make([][]rune, len(lines))
	for i, line := range lines {
		matrix[i] = []rune(line)
	}

	return matrix, nil
}

func isXmas(word string) bool {
	return word == "XMAS" || word == "SAMX"
}

func isMas(word string) bool {
	return word == "MAS" || word == "SAM"
}

func countXmas(matrix [][]rune) int {
	rowCount := len(matrix)
	colCount := len(matrix[0])
	var count int

	for rowIndex := 0; rowIndex < rowCount; rowIndex++ {
		for colIndex := 0; colIndex < colCount; colIndex++ {
			words := [4]string{}

			// Horizontal word
			if colIndex < colCount-3 {
				words[0] = string(matrix[rowIndex][colIndex : colIndex+4])
			}

			// Vertical word
			if rowIndex < rowCount-3 {
				for i := 0; i <= 3; i++ {
					words[1] += string(matrix[rowIndex+i][colIndex])
				}
			}

			// Diagonal words
			if rowIndex < rowCount-3 && colIndex < colCount-3 {
				for i := 0; i <= 3; i++ {
					words[2] += string(matrix[rowIndex+i][colIndex+i])
					words[3] += string(matrix[rowIndex+(3-i)][colIndex+i])
				}
			}

			for _, word := range words {
				if isXmas(word) {
					count++
				}
			}
		}
	}

	return count
}

func countCrossMas(matrix [][]rune) int {
	rowCount := len(matrix)
	colCount := len(matrix[0])
	var count int

	for rowIndex := 0; rowIndex < rowCount-2; rowIndex++ {
		for colIndex := 0; colIndex < colCount-2; colIndex++ {
			var downDiagonalWord, upDiagonalWord string

			for i := 0; i <= 2; i++ {
				downDiagonalWord += string(matrix[rowIndex+i][colIndex+i])
				upDiagonalWord += string(matrix[rowIndex+(2-i)][colIndex+i])
			}

			if isMas(downDiagonalWord) && isMas(upDiagonalWord) {
				count++
			}
		}
	}

	return count
}

func main() {
	matrix, err := readInput()
	if err != nil {
		log.Fatalln(err)
	}

	// Part one
	fmt.Println("Number of XMAS occurrences:", countXmas(matrix))
	// Part two
	fmt.Println("Number of X-MAS occurrences:", countCrossMas(matrix))
}
